src/analysis/normed/group/pointwise.lean

```diff
@@ -3,16 +3,117 @@ Copyright (c) 2021 Sébastien Gouëzel. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Sébastien Gouëzel
 -/
-import analysis.normed.group.basic
+import analysis.convex.basic
+import analysis.normed.group.add_torsor
 import topology.metric_space.hausdorff_distance
 
 /-!
-# Properties of pointwise addition of sets in normed groups.
+# Properties of pointwise addition of sets in normed groups
 
 We explore the relationships between pointwise addition of sets in normed groups, and the norm.
 Notably, we show that the sum of bounded sets remain bounded.
 -/
 
+section
+variables {α : Type*} [ordered_semiring α] {a b : α}
+
+lemma mul_lt_of_lt_one_left (hb : 0 < b) (ha : a < 1) : a * b < b :=
+(mul_lt_mul_of_pos_right ha hb).trans_le (one_mul _).le
+
+lemma mul_lt_of_lt_one_right (ha : 0 < a) (hb : b < 1) : a * b < a :=
+(mul_lt_mul_of_pos_left hb ha).trans_le (mul_one _).le
+
+end
+
+namespace emetric
+variables {α : Type*} [pseudo_emetric_space α] {δ ε : ℝ} {s : set α} {x y : α}
+
+lemma inf_edist_le_edist_add_inf_edist : inf_edist x s ≤ edist x y + inf_edist y s :=
+by { rw add_comm, exact inf_edist_le_inf_edist_add_edist }
+
+end emetric
+
+namespace metric
+variables {α : Type*} [pseudo_emetric_space α] {δ ε : ℝ} {s : set α} {x : α}
+
+open emetric set
+
+alias ennreal.of_real_eq_zero ↔ _ ennreal.of_real_of_nonpos
+
+lemma mem_cthickening_iff : x ∈ cthickening δ s ↔ inf_edist x s ≤ ennreal.of_real δ := iff.rfl
+
+lemma thickening_of_nonpos (hδ : δ ≤ 0) (s : set α) : thickening δ s = ∅ :=
+eq_empty_of_subset_empty $ λ x hx, (ennreal.of_real_of_nonpos hδ).not_gt $ bot_le.trans_lt hx
+
+@[simp] lemma cthickening_closure : cthickening δ (closure s) = cthickening δ s :=
+by simp_rw [cthickening, inf_edist_closure]
+
+@[simp] lemma thickening_thickening_subset (ε δ : ℝ) (s : set α) :
+  thickening ε (thickening δ s) ⊆ thickening (ε + δ) s :=
+begin
+  obtain hε | hε := le_total ε 0,
+  { simp only [thickening_of_nonpos hε, empty_subset] },
+  obtain hδ | hδ := le_total δ 0,
+  { simp only [thickening_of_nonpos hδ, thickening_empty, empty_subset] },
+  intros x,
+  simp_rw [mem_thickening_iff_exists_edist_lt, ennreal.of_real_add hε hδ],
+  exact λ ⟨y, ⟨z, hz, hy⟩, hx⟩, ⟨z, hz, (edist_triangle _ _ _).trans_lt $ ennreal.add_lt_add hx hy⟩,
+end
+
+@[simp] lemma thickening_cthickening_subset (ε : ℝ) (hδ : 0 ≤ δ) (s : set α) :
+  thickening ε (cthickening δ s) ⊆ thickening (ε + δ) s :=
+begin
+  obtain hε | hε := le_total ε 0,
+  { simp only [thickening_of_nonpos hε, empty_subset] },
+  intro x,
+  simp_rw [mem_thickening_iff_exists_edist_lt, mem_cthickening_iff, ←inf_edist_lt_iff,
+    ennreal.of_real_add hε hδ],
+  rintro ⟨y, hy, hxy⟩,
+  refine inf_edist_le_edist_add_inf_edist.trans_lt sorry,
+  sorry -- done by `with_top.add_lt_add_of_lt_of_le`
```

Comment:
There's an awaiting review tag on this PR, but in fact I'm not sure it's ready for review :-)

Reply:
#13380 takes care of that one :)